hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java

```diff
@@ -1521,8 +1547,8 @@ private void deleteInternal(Container container, boolean force)
       container.writeUnlock();
     }
     // Avoid holding write locks for disk operations
-    container.delete();
     sendICR(container);
+    container.delete();
```

Comment:
If the delete operation fails, does it cause a resource leak, since the ICR have sent to SCM, then the SCM will not retry to delete the Container again?

Reply:
Even with the current code, the container is out of the container map, and not able to be found by a subsequent delete operation. The original code would fail to send the ICR and fail to process further deletes if the delete fails. The container data will have been moved into the tmp folder before the ICR has been sent and can be cleaned up by other means. Is there anything that cleans this folder on startup to deal with failures on the current code path?

Reply:
> Is there anything that cleans this folder on startup to deal with failures on the current code path?

Yes tmp dir should be cleaned on startup and graceful shutdown to handle cases where the original delete from the tmp dir failed. See callers of [this method](https://github.com/apache/ozone/blob/4e603aa93c8479349776c1597ef54504453cb512/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/HddsVolume.java#L214). We just need to make sure the container is moved to the tmp dir and DB entries are deleted before acking the delete back to SCM, which the current version of this patch handles.